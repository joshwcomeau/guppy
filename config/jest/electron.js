// @flow
global.window = {
  require: require,
};
