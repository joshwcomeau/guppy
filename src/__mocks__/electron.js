/* eslint-disable flowtype/require-valid-file-annotation */
import * as path from 'path';

module.exports = {
  ipcRenderer: {
    send: jest.fn(),
  },
  remote: {
    app: {
      getAppPath: () => path.resolve(__dirname, '..', '..', '..'),
      getPath: () =>
        process.env.APPDATA ||
        (process.platform === 'darwin'
          ? process.env.HOME + 'Library/Preferences'
          : '/var/local'),
    },
    dialog: {
      showErrorBox: jest.fn(),
      showOpenDialog: jest.fn(),
      showMessageBox: jest.fn(),
    },
    shell: {
      moveItemToTrash: jest.fn(),
    },
  },
};
