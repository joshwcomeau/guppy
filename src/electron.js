/**
 * This is our main Electron process.
 * It handles opening our app window, and quitting the application.
 */
const { app, BrowserWindow, ipcMain, dialog } = require('electron');
const { autoUpdater } = require('electron-updater');
const log = require('electron-log');
const path = require('path');
const url = require('url');
const fixPath = require('fix-path');
const chalkRaw = require('chalk');

const killProcessId = require('./services/kill-process-id.service');
const electronStore = require('./services/electron-store.service');

const chalk = new chalkRaw.constructor({ level: 3 });
let icon256Path = '../public/256x256.png';

// In production, we need to use `fixPath` to let Guppy use NPM.
// For reasons unknown, the opposite is true in development; adding this breaks
// everything.
if (process.env.NODE_ENV !== 'development') {
  icon256Path = '../build/256x256.png';
  fixPath();
}

// Keep a global reference of the window object, if you don't, the window will
// be closed automatically when the JavaScript object is garbage collected.
let mainWindow;

// Guppy is primarily a process runner, and so many processes may be started.
// To ensure that Guppy cleans up after itself, we'll store all of the spawned
// process IDs up here in the main process, so that we can terminate them all
// before the app quits.
let processIds = [];

// Used for renderer tasks on close (emitting event app-will-close once)
let emitAppWillClose = true;

const MOVE_TO_APP_FOLDER_KEY = 'leave-application-in-original-location';

// This logging setup is not required for auto-updates to work,
// but it sure makes debugging easier :)
//-------------------------------------------------------------------
autoUpdater.logger = log;
autoUpdater.logger.transports.file.level = 'info';

// default logs from electron-updater are
// 'Checking for update', 'Found verison ...', 'Downloading update ...'

autoUpdater.on('update-downloaded', (event, releaseNotes, releaseName) => {
  const dialogOpts = {
    type: 'info',
    buttons: ['Restart', 'Later'],
    title: 'Application Update',
    message: process.platform === 'win32' ? releaseNotes : releaseName,
    detail:
      'A new version has been downloaded. Restart the application to apply the updates.',
  };

  dialog.showMessageBox(dialogOpts, response => {
    if (response === 0) autoUpdater.quitAndInstall();
  });
});

function createWindow() {
  // Check if there is a newer version
  autoUpdater.checkForUpdatesAndNotify();

  // Verifies if Guppy is already in the Applications folder
  // and prompts the user to move it if it isn't
  manageApplicationLocation();

  // Create the browser window.
  mainWindow = new BrowserWindow({
    width: 1120,
    height: 768,
    minWidth: 777,
    titleBarStyle: 'hidden',
    icon: path.join(__dirname, icon256Path),
  });

  // set up some chrome extensions
  if (process.env.NODE_ENV === 'development') {
    const {
      default: installExtension,
      REACT_DEVELOPER_TOOLS,
      REACT_PERF,
    } = require('electron-devtools-installer');

    require('electron-debug')({ showDevTools: true });

    const ChromeLens = {
      // ID of the extension (https://chrome.google.com/webstore/detail/chromelens/idikgljglpfilbhaboonnpnnincjhjkd)
      id: 'idikgljglpfilbhaboonnpnnincjhjkd',
      electron: '>=1.2.1',
    };

    const extensions = [REACT_DEVELOPER_TOOLS, REACT_PERF, ChromeLens];

    for (const extension of extensions) {
      try {
        installExtension(extension);
      } catch (e) {
        console.error(chalk.red(`[ELECTRON] Extension installation failed`), e);
      }
    }
  }

  // and load the index.html of the app.
  const startUrl =
    process.env.ELECTRON_START_URL ||
    url.format({
      pathname: path.join(__dirname, '/../build/index.html'),
      protocol: 'file:',
      slashes: true,
    });
  mainWindow.loadURL(startUrl);

  mainWindow.on('close', e => {
    if (emitAppWillClose) {
      emitAppWillClose = false;
      mainWindow.webContents.send('app-will-close');
      e.preventDefault();
    }
  });

  // Emitted when the window is closed.
  mainWindow.on('closed', function() {
    // Dereference the window object, usually you would store windows
    // in an array if your app supports multi windows, this is the time
    // when you should delete the corresponding element.
    mainWindow = null;
  });
}

// This method will be called when Electron has finished
// initialization and is ready to create browser windows.
// Some APIs can only be used after this event occurs.
app.on('ready', createWindow);

// Quit when all windows are closed.
app.on('window-all-closed', function() {
  // On OS X it is common for applications and their menu bar
  // to stay active until the user quits explicitly with Cmd + Q
  if (process.platform !== 'darwin') {
    app.quit();
  }
});

app.on('activate', function() {
  // On OS X it's common to re-create a window in the app when the
  // dock icon is clicked and there are no other windows open.
  if (mainWindow === null) {
    createWindow();
  }
});

ipcMain.on('addProcessId', (event, processId) => {
  processIds.push(processId);
});

ipcMain.on('removeProcessId', (event, processId) => {
  processIds = processIds.filter(id => id !== processId);
});

ipcMain.on('killAllRunningProcesses', event => {
  if (processIds.length) {
    event.preventDefault();
    killAllRunningProcesses();
  }
  app.quit();
});

ipcMain.on('triggerClose', (e, proceed) => {
  if (!proceed) {
    // user aborted
    emitAppWillClose = true; // reset flag
    return;
  }

  mainWindow.close();
});

const killAllRunningProcesses = () => {
  try {
    processIds.forEach(processId => {
      killProcessId(processId);

      // Remove the parent or any children PIDs from the list of tracked
      // IDs, since they're killed now.
      processIds = processIds.filter(id => id !== processId);
    });
  } catch (err) {
    console.error('Got error when trying to kill children', err);
  }
};

const canApplicationBeMoved = () => {
  // The application can be moved if :
  //  - The platform is MacOS
  //  - The function 'isInApplicationsFolder' exists
  //  - The app is running in production
  //  - Guppy is not already in the Applications folder
  //  - The user hasn't chosen to not see the dialog prompt again
  const hasApplicationsFolder =
    process.platform === 'darwin' &&
    typeof app.isInApplicationsFolder === 'function';

  // NOTE: Because this file isn't compiled by Webpack, `process.env.NODE_ENV`
  // will be undefined in the bundled application. Rather than check to see if
  // it's set to `production`, we just care that it ISN'T set to `development`
  // (development is set in package.json when running the 'start' script).
  const isProduction = process.env.NODE_ENV !== 'development';

  if (
    hasApplicationsFolder &&
    isProduction &&
    !app.isInApplicationsFolder() &&
    !electronStore.has(MOVE_TO_APP_FOLDER_KEY)
  ) {
    return true;
  }

  return false;
};

const manageApplicationLocation = () => {
  if (canApplicationBeMoved()) {
    dialog.showMessageBox(
      {
        type: 'question',
        buttons: ['Yes, move to Applications', 'No thanks'],
        message: 'Move to Applications folder?',
        detail:
          "I see that I'm not in the Applications folder. I can move myself there if you'd like!",
        icon: path.join(__dirname, icon256Path),
        cancelId: 1,
        defaultId: 0,
        checkboxLabel: 'Do not show this message again',
      },
      (res, checkboxChecked) => {
        // User wants the application to be moved
        if (res === 0) {
          try {
            app.moveToApplicationsFolder();
          } catch (err) {
            dialog.showErrorBox(
              'Error',
              'Could not move Guppy to the Applications folder'
            );
            console.error(
              'Could not move Guppy to the Applications folder',
              err
            );
          }
        }

        // User doesn't want to see the message again
        if (checkboxChecked === true) {
          electronStore.set(MOVE_TO_APP_FOLDER_KEY, true);
        }
      }
    );
  }
};
