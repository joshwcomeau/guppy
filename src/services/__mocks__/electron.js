import * as path from 'path';
module.exports = {
  remote: {
    app: {
      getAppPath: () => path.resolve(__dirname, '..', '..', '..'),
      getPath: () =>
        process.env.APPDATA ||
        (process.platform == 'darwin'
          ? process.env.HOME + 'Library/Preferences'
          : '/var/local'),
    },
  },
};
