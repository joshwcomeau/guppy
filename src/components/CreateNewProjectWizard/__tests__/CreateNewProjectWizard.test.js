import React from 'react';
import { shallow } from 'enzyme';
import { remote } from 'electron'; // Mocked

import {
  CreateNewProjectWizard,
  dialogOptionsFolderExists,
  dialogCallbackFolderExists,
  dialogStarterNotFoundErrorArgs,
  FORM_STEPS,
} from '../CreateNewProjectWizard';
import { initialState as appSettingsInitialState } from '../../../reducers/app-settings.reducer';

const projectHomePath = '/users/guppy-dev';

jest.mock('../../../services/create-project.service', () => ({
  checkIfProjectExists: jest.fn(
    (path, name) => path === '/users/guppy-dev' && name === 'first-project'
  ),
  getProjectNameSlug: jest.requireActual(
    '../../../services/create-project.service'
  ).getProjectNameSlug,
}));

jest.mock('../../../services/check-if-url-exists.service', () => ({
  urlExists: jest.fn(starterUrl =>
    Promise.resolve(starterUrl.includes('gatsby-blog-starter'))
  ),
}));

jest.useFakeTimers();

const { dialog } = remote;

describe('CreateNewProjectWizard component', () => {
  let wrapper;
  let instance;
  let mockActions;

  // Create some projects with just a name key as it is needed for uniqueness check
  const mockProjects = {
    'first-project-id': {
      name: 'first-project',
    },
    'second-project-id': {
      name: 'second-project',
    },
  };

  const newProject = {
    name: 'New project',
    projectType: 'create-react-app',
    projectIcon: 'icon',
  };

  const mockAppSettingsState = {
    ...appSettingsInitialState,
    general: {
      defaultProjectPath: projectHomePath,
    },
  };

  beforeEach(() => {
    mockActions = {
      addProject: jest.fn(),
      createNewProjectCancel: jest.fn(),
      createNewProjectFinish: jest.fn(),
    };
    window.clearTimeout = jest.fn();
    wrapper = shallow(
      <CreateNewProjectWizard
        projects={mockProjects}
        projectHomePath={projectHomePath}
        isVisible={true}
        isOnboardingCompleted={true}
        settings={mockAppSettingsState}
        {...mockActions}
      />
    );
    instance = wrapper.instance();
  });
  it('should render TwoPaneModal', () => {
    expect(wrapper.renderProp('children')(true)).toMatchSnapshot();
  });

  it('should initialize on mount', () => {
    instance.reinitialize = jest.fn();
    instance.componentDidMount();
    expect(instance.reinitialize).toHaveBeenCalled();
  });

  it('should clear timeout on unmount', () => {
    instance.componentWillUnmount();
    expect(window.clearTimeout).toHaveBeenCalledWith(instance.timeoutId);
  });

  it('should update field and check uniqueness of projectName', () => {
    const newProjectName = 'New Project Name';
    instance.verifyProjectNameUniqueness = jest.fn();
    instance.updateFieldValue('projectName', newProjectName);
    expect(instance.state.projectName).toEqual(newProjectName);
    expect(instance.state.activeField).toEqual('projectName');
    expect(instance.verifyProjectNameUniqueness).toHaveBeenCalledWith(
      newProjectName
    );

    // update other field
    instance.verifyProjectNameUniqueness = jest.fn();
    instance.updateFieldValue('projectIcon', 'new-icon');
    expect(instance.verifyProjectNameUniqueness).not.toHaveBeenCalled();
  });

  it('should set focus', () => {
    instance.focusField('projectName');
    expect(instance.state.activeField).toEqual('projectName');
  });

  it('should update isProjectTaken', () => {
    instance.updateFieldValue('projectName', 'First project');
    expect(instance.state.isProjectNameTaken).toBe(true);
    instance.updateFieldValue('projectName', 'New project');
    expect(instance.state.isProjectNameTaken).toBe(false);
  });

  it('should check project location usage (exists on disk)', () => {
    wrapper.setState({
      projectName: 'first-project',
    });
    instance.checkProjectLocationUsage();
    expect(dialog.showMessageBox).toHaveBeenCalledWith(
      dialogOptionsFolderExists,
      expect.anything()
    );

    const resolveMock = jest.fn();
    const rejectMock = jest.fn();
    dialogCallbackFolderExists(resolveMock, rejectMock).call(instance, 0);
    expect(rejectMock).toHaveBeenCalled();

    dialogCallbackFolderExists(resolveMock, rejectMock).call(instance, 1);
    expect(resolveMock).toHaveBeenCalled();
  });

  it('should check project location usage (new project)', async () => {
    wrapper.setState({
      projectName: 'new-project',
    });
    await expect(instance.checkProjectLocationUsage()).resolves.toBeUndefined();
  });

  it('should check if starter url exists', async () => {
    wrapper.setState({
      projectStarter: 'gatsby-blog-starter',
    });

    expect(instance.checkIfStarterUrlExists()).resolves.not.toThrow();

    // should return undefined for empty starter
    wrapper.setState({
      projectStarter: '',
    });

    expect(await instance.checkIfStarterUrlExists()).toBeUndefined();

    wrapper.setState({
      projectStarter: 'not-existing-starter',
    });

    await expect(instance.checkIfStarterUrlExists()).rejects.toThrow(
      'starter-not-found'
    );

    expect(dialog.showErrorBox).toHaveBeenCalledWith(
      ...dialogStarterNotFoundErrorArgs(
        'https://github.com/gatsbyjs/not-existing-starter'
      )
    );
  });

  it('should dispatch createNewProjectFinish & add project (finishBuilding)', () => {
    const projectType = 'create-react-app';
    wrapper.setState({
      projectType,
    });
    instance.reinitialize = jest.fn();

    instance.finishBuilding(newProject);
    expect(mockActions.createNewProjectFinish).toHaveBeenCalled();

    jest.runAllTimers();
    expect(mockActions.addProject).toHaveBeenCalledWith(
      newProject,
      projectHomePath,
      projectType,
      true
    );
    expect(instance.reinitialize).toHaveBeenCalled();
  });

  it('should throw error if projectType not defined (finishBuilding)', () => {
    expect(instance.finishBuilding).toThrow(
      'Project created without projectType'
    );
  });

  describe('Submit button (next step & build)', () => {
    it('should set next step if not build click', () => {
      wrapper.setState({
        currentStep: 'projectName',
        projectType: 'gatsby',
      });
      instance.handleSubmit();
      expect(instance.state).toEqual(
        expect.objectContaining({
          currentStep: FORM_STEPS[1],
          activeField: FORM_STEPS[1],
        })
      );
    });

    it('should trigger build & checks before starting build', async () => {
      // mock checks
      instance.checkProjectLocationUsage = jest.fn();
      instance.checkIfStarterUrlExists = jest.fn();
      wrapper.setState({
        currentStep: 'projectStarter',
        projectType: 'gatsby',
      });
      await instance.handleSubmit();
      expect(instance.checkProjectLocationUsage).toHaveBeenCalled();
      expect(instance.checkIfStarterUrlExists).toHaveBeenCalled();
      expect(instance.state.status).toEqual('building-project');
    });
  });
});
