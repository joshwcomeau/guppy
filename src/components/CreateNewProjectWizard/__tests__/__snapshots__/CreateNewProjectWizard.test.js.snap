// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateNewProjectWizard component should render TwoPaneModal 1`] = `
<Fragment>
  <TwoPaneModal
    backface={
      <BuildPane
        handleCompleteBuild={[Function]}
        projectHomePath="/users/guppy-dev"
        projectIcon={null}
        projectName=""
        projectStarter=""
        status="filling-in-form"
      />
    }
    isDismissable={true}
    isFolded={false}
    leftPane={
      <Debounced
        duration={40}
        on="projectName"
      >
        <SummaryPane
          activeField="projectName"
          currentStep="projectName"
        />
      </Debounced>
    }
    onDismiss={[MockFunction]}
    rightPane={
      <MainPane
        activeField="projectName"
        currentStepIndex={0}
        focusField={[Function]}
        handleSubmit={[Function]}
        hasBeenSubmitted={false}
        isProjectNameTaken={false}
        projectIcon={null}
        projectName=""
        projectStarter=""
        status="filling-in-form"
        updateFieldValue={[Function]}
      />
    }
    transitionState={true}
  />
</Fragment>
`;
